src/ALE/PLM_functions.F90

```diff
@@ -5,21 +5,177 @@ module PLM_functions
+  endif
+
+  ! An attempt to avoid inconsistency when the values become unrepresentable.
+  if (abs(PLM_slope_wa) < 1.E-140) PLM_slope_wa = 0.
+
+end function PLM_slope_wa
+
+!> Returns a limited PLM slope following Colella and Woodward 1984.
+real elemental pure function PLM_slope_cw(h_l, h_c, h_r, h_neglect, u_l, u_c, u_r)
+  real, intent(in) :: h_l !< Thickness of left cell [units of grid thickness]
+  real, intent(in) :: h_c !< Thickness of center cell [units of grid thickness]
+  real, intent(in) :: h_r !< Thickness of right cell [units of grid thickness]
+  real, intent(in) :: h_neglect !< A negligible thickness [units of grid thickness]
+  real, intent(in) :: u_l !< Value of left cell [units of u]
+  real, intent(in) :: u_c !< Value of center cell [units of u]
+  real, intent(in) :: u_r !< Value of right cell [units of u]
+  ! Local variables
+  real :: sigma_l, sigma_c, sigma_r, u_min, u_max, h_cn
+
+  h_cn = h_c + h_neglect
+
+  ! Side differences
+  sigma_r = u_r - u_c
+  sigma_l = u_c - u_l
+
+  ! This is the second order slope given by equation 1.7 of
+  ! Piecewise Parabolic Method, Colella and Woodward (1984),
+  ! http://dx.doi.org/10.1016/0021-991(84)90143-8.
+  ! For uniform resolution it simplifies to ( u_r - u_l )/2 .
+  sigma_c = ( h_c / ( h_cn + ( h_l + h_r ) ) ) * ( &
+                ( 2.*h_l + h_c ) / ( h_r + h_cn ) * sigma_r &
+              + ( 2.*h_r + h_c ) / ( h_l + h_cn ) * sigma_l )
+
+  ! Limit slope so that reconstructions are bounded by neighbors
+  u_min = min( u_l, u_c, u_r )
+  u_max = max( u_l, u_c, u_r )
+  if ( (sigma_l * sigma_r) > 0.0 ) then
+    ! This limits the slope so that the edge values are bounded by the
+    ! two cell averages spanning the edge.
+    PLM_slope_cw = sign( min( abs(sigma_c), 2.*min( u_c - u_min, u_max - u_c ) ), sigma_c )
+  else
+    ! Extrema in the mean values require a PCM reconstruction avoid generating
+    ! larger extreme values.
+    PLM_slope_cw = 0.0
+  endif
+
+  ! This block tests to see if roundoff causes edge values to be out of bounds
+  if (u_c - 0.5*abs(PLM_slope_cw) < u_min .or.  u_c + 0.5*abs(PLM_slope_cw) > u_max) then
+    PLM_slope_cw = PLM_slope_cw * ( 1. - epsilon(PLM_slope_cw) )
+  endif
+
+  ! An attempt to avoid inconsistency when the values become unrepresentable.
+  if (abs(PLM_slope_cw) < 1.E-140) PLM_slope_cw = 0.
+
+end function PLM_slope_cw
+
+!> Returns a limited PLM slope following Colella and Woodward 1984.
+real elemental pure function PLM_monotonized_slope(u_l, u_c, u_r, s_l, s_c, s_r)
+  real, intent(in) :: u_l !< Value of left cell [units of u]
+  real, intent(in) :: u_c !< Value of center cell [units of u]
+  real, intent(in) :: u_r !< Value of right cell [units of u]
+  real, intent(in) :: s_l !< PLM slope of left cell [units of u]
+  real, intent(in) :: s_c !< PLM slope of center cell [units of u]
+  real, intent(in) :: s_r !< PLM slope of right cell [units of u]
+  ! Local variables
+  real :: e_r, e_l, edge, almost_two, slp
```

Comment:
Please document the units of these variables.

Reply:
Addressed in 553166ae8e8a871a0d1e39a091cbe707ef8a3159